staging/src/k8s.io/apimachinery/pkg/watch/streamwatcher_test.go

```diff
@@ -105,3 +139,77 @@ func TestStreamWatcherError(t *testing.T) {
 		t.Fatalf("unexpected open channel")
 	}
 }
+
+func TestStreamWatcherStopRace(t *testing.T) {
+	table := []Event{
+		{Type: Added, Object: testType("foo")},
+		{Type: Added, Object: testType("bar")},
+	}
+
+	fd := &fakeDecoder{items: make(chan Event, 5)}
+	fr := &fakeReporter{}
+	sw := NewStreamWatcher(fd, fr)
+
+	for i, item := range table {
+		fd.items <- item
+		fmt.Printf("sent event %d: %+v\n", i+1, item)
+	}
+
+	fmt.Printf("waiting for event 1\n")
+	got, open := <-sw.ResultChan()
+	if !open {
+		t.Errorf("unexpected early close")
+	}
+	if e, a := table[0], got; !reflect.DeepEqual(e, a) {
+		t.Errorf("expected %v, got %v", e, a)
+	} else {
+		fmt.Printf("GOT 1: %+v\n", got)
+	}
+
+	for fd.getCount() < 2 {
+		fmt.Printf("waiting for receive...\n")
+		time.Sleep(time.Second)
```

Comment:
I can see why you're doing it this way but I worry that this is inherently flaky, especially if CI tests are under heavy load. I think personally I would have gone for a randomized approach, where it just tries it a bunch of times from parallel goroutines and verifies that every result channel actually gets closed. Test on the unfixed code to figure out how many times we have to try to reliably trigger the issue.

(That style of test might also be flaky, but in the other direction-- it might give false passes instead of false failures. But that's OK because if the code really does break, it'll show up as a flaky test, which is OK when there's actually a problem that needs to be fixed.)

Anyway, I'm not sure if this is mandatory, I'm not sure this test would flake very often. Let's see what @wojtek-t thinks.

Reply:
Maybe instead of sleeping call goruntime.Gosched() a few times?

Reply:
For sure, Gosched does the job. I'll fix this.

Reply:
The send and synchronization is exactly done in a way, so that the dedicated race condition is met. If this is possible I would avoid to use some brute force test that only fails with a certain probability.

Reply:
OK, if this is reliable I'm willing to give it a shot.

But just for the record things that happen "with a certain probability" can actually be more certain overall, because they're more robust to external interference from e.g. CPU contention. (Relevant background: https://en.wikipedia.org/wiki/Central_limit_theorem). The hand-tuned version of things like this can be brittle in surprising ways.

That said since you already have written this, let's give it a shot and see if it turns out to be a problem in practice. I just want to plug the brute force methods for next time, they may not be elegant looking, but they work...

Reply:
OK actually I just pulled this and checked to see if it is flaky:

```
$ go test k8s.io/apimachinery/pkg/watch/ --count=10000 | grep "FAIL: TestStreamWatcher" | wc -l
1240
```

That's a 12% flake rate, seems to go up if I add --race. I think I can't knowingly merge something with a flake rate higher than .01%

Reply:
running with -race under [stress](https://pkg.go.dev/golang.org/x/tools/cmd/stress), the failure rate is 40%:

```
go test -race -c k8s.io/apimachinery/pkg/watch
stress ./watch.test -test.run TestStreamWatcherStopRace
...
1744 runs so far, 698 failures
```

Reply:
I've played with various combinations, but I did not make it to reliably enforce the undesired blocking channel send.

Any ideas how to write a test for this are welcome.

Reply:
I think you have to set up a condition where two goroutines are racing and run it a bunch of times, like I said :/

Reply:
Ok, another variant. This one is unfortunately more invasive, because it requires a test case handling in the code (just an if)
For the error case 10000 retries showed 100% failures, and the good case only 0.0003% (because of the final wait time limitation). May be this is better. But now I don't have any other idea anymore.